### PR TITLE
Changes from yacademy review

### PR DESCRIPTION
**Pull request by simonbdefi**

### CODE REVIEW

File: interfaces/Stargate/IPool.sol [hunk 1]
```diff
@@ -1,16 +1,12 @@
 // SPDX-License-Identifier: BUSL-1.1
-
 pragma solidity 0.8.15;
 
 interface IPool {
-    function poolId() external view returns (uint256); // shared id between chains to represent same pool
-    function token() external view returns (address); // the token for the pool
-    function router() external view returns (address); // the router for the pool
-    function totalLiquidity() external view returns (uint256); // the total amount of tokens added on this side of the chain (fees + deposits - withdrawals)
```
Comment (newmickymousse) [inline interfaces/Stargate/IPool.sol:9]: Why remove the comments or meanings of the functions?

File: tests/Mainnet/test_redeem_local.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    strategy.harvest({"from": gov})
+    assert pytest.approx(strategy.estimatedTotalAssets(), rel=RELATIVE_APPROX) == amount
+
+    # whale withdraws all available deltaCredit
+    liquidityPool = Contract(strategy.liquidityPool())
+    router = Contract(liquidityPool.router())
+    router.instantRedeemLocal(liquidityPool.poolId(), liquidityPool.deltaCredit(), strategist, {"from":token_LP_whale})
+    assert liquidityPool.deltaCredit() < amount
+
+    # unstake all lp
```
Comment (newmickymousse) [inline tests/Mainnet/test_redeem_local.py:25]: I would add an assert to check that balanceOfStakedLPToken > 0

File: tests/Mainnet/test_redeem_local.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    strategy.unstakeLP(strategy.balanceOfStakedLPToken(), {"from": gov})
+    assert strategy.balanceOfStakedLPToken() == 0
```
Comment (newmickymousse) [inline tests/Mainnet/test_redeem_local.py:29]: I would add an assert to check that balanceOfUnstakedLPToken > 0

File: tests/Mainnet/test_redeem_local.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    
+    # USDT - pull from Arbitrum
+    if strategy.want() == "0xdAC17F958D2ee523a2206206994597C13D831ec7":
+        tx = strategy.redeemLocal(110, 2, strategy.balanceOfUnstakedLPToken(), {"from": gov, "amount":1e17})
+    
+    # WETH - Pull from Optimism
+    if strategy.want() == "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2":
+        tx = strategy.redeemLocal(111, 13, strategy.balanceOfUnstakedLPToken(), {"from": gov, "amount":1e17})
+    
+    assert strategy.balanceOfUnstakedLPToken() == 0 # make sure we have transferred all LP tokens using redeemLocal
```
Comment (newmickymousse) [inline tests/Mainnet/test_redeem_local.py:45]: How long after the redeemLocal does it take to receive the want?

File: tests/Optimism/test_redeem_local.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import brownie
```
Comment (newmickymousse) [inline tests/Optimism/test_redeem_local.py:1]: Please, check other comments in the other test

**Review by newmickymousse — CHANGES_REQUESTED** [review]

File: interfaces/Stargate/IPool.sol (hunk 1 above)
Comment (simonbdefi) [reply interfaces/Stargate/IPool.sol:9]: reinstated

File: tests/Mainnet/test_redeem_local.py (hunk 2 above)
Comment (simonbdefi) [reply tests/Mainnet/test_redeem_local.py:25]: Changed for both mainnet and opti tests

File: tests/Mainnet/test_redeem_local.py (hunk 3 above)
Comment (simonbdefi) [reply tests/Mainnet/test_redeem_local.py:29]: Changed for both mainnet and opti tests

File: tests/Optimism/test_redeem_local.py (hunk 5 above)
Comment (simonbdefi) [reply tests/Optimism/test_redeem_local.py:1]: done!

File: tests/Mainnet/test_redeem_local.py (hunk 4 above)
Comment (simonbdefi) [reply tests/Mainnet/test_redeem_local.py:45]: RedeemLocal is block confs of source chain plus block confs of B chain, so less than 1 hour.

Comment (simonbdefi) [reply tests/Mainnet/test_redeem_local.py:45]: Worst case would be with polygon (512 blocks, about 20 minutes) + other chain (i.e. ETH, about 3-4 min).

**Comment by simonbdefi** [comment]
Pasting the suggested fix here to have it visible for yacademy who audited Stargate V2,

> > https://github.com/0xValJohn/yearn-stargate/blob/ba019f52473ee739a16072e003b530e1dee606db/contracts/Strategy.sol#L237-L238 Critic! Why only compare the _liquidatedAmount with amountNeeded? You should compare (balanceOfWant() > _liquidatedAmount) >  > 
> > Scenario: >  amountNeeded = 10 preWithdrawnWant = 8 balanceOfAllLPTokenStakedInWant = 8 balanceOfAllLPTokenIdleInWant = 0
> > Now you liquidated 2 from balanceOfAllLPTokenStakedInWant now: >  >  balanceOfWant() = 10  > balanceOfAllLPTokenStakedInWant = 6
> > When you come to this line you’ll do this: > 10>2 ? Go inside if  > balanceOfLPTokens = 6 potentialLoss = 10-2 = 8
> > Loss = 8 > 6 ? 8-6 : 0 Which loss = 2 Although we have the want needed in the first place.
> 
Suggest the following:
 
 ```solidity
 if (_amountNeeded > _balanceOfWant) {
     uint256 balanceOfLPTokens = _lpToLd(balanceOfAllLPToken());
     uint256 _potentialLoss = _amountNeeded - _liquidatedAmount;
     unchecked {
         _loss = _potentialLoss > balanceOfLPTokens ? _potentialLoss - balanceOfLPTokens : 0;
     }
 }
 ```
 
 Would like to get vetting from all you 3 reviewers (@mil0xeth @tapired @newmickymousse) as this function is critical.
